pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -529,29 +530,36 @@ func (s *Scheduler) addToNewNodeClaim(ctx context.Context, pod *corev1.Pod) erro
 	var offeringsToReserve []*cloudprovider.Offering
 
+	// If AutoRelaxMinValues is enabled, we go through the nodepools twice - first with the fallback disabled and then with it enabled since we only want to relax the min values if there's no nodepool
+	// which the pod will schedule against without relaxing.
+	queueLength := lo.Ternary(karpopts.FromContext(ctx).FeatureGates.AutoRelaxMinValues, len(s.nodeClaimTemplates)*2, len(s.nodeClaimTemplates))
+
 	errs := make([]error, len(s.nodeClaimTemplates))
-	parallelizeUntil(s.numConcurrentReconciles, len(s.nodeClaimTemplates), func(i int) bool {
-		its := s.nodeClaimTemplates[i].InstanceTypeOptions
+	parallelizeUntil(s.numConcurrentReconciles, queueLength, func(i int) bool {
+		nodeClaimTemplateIndex := i % len(s.nodeClaimTemplates)
+		its := s.nodeClaimTemplates[nodeClaimTemplateIndex].InstanceTypeOptions
 		// if limits have been applied to the nodepool, ensure we filter instance types to avoid violating those limits
-		if remaining, ok := s.remainingResources[s.nodeClaimTemplates[i].NodePoolName]; ok {
+		if remaining, ok := s.remainingResources[s.nodeClaimTemplates[nodeClaimTemplateIndex].NodePoolName]; ok {
 			its = filterByRemainingResources(its, remaining)
 			if len(its) == 0 {
-				errs[i] = serrors.Wrap(fmt.Errorf("all available instance types exceed limits for nodepool"), "NodePool", klog.KRef("", s.nodeClaimTemplates[i].NodePoolName))
+				errs[nodeClaimTemplateIndex] = serrors.Wrap(fmt.Errorf("all available instance types exceed limits for nodepool"), "NodePool", klog.KRef("", s.nodeClaimTemplates[nodeClaimTemplateIndex].NodePoolName))
 				return true
-			} else if len(s.nodeClaimTemplates[i].InstanceTypeOptions) != len(its) {
+			} else if len(s.nodeClaimTemplates[nodeClaimTemplateIndex].InstanceTypeOptions) != len(its) {
 				log.FromContext(ctx).V(1).WithValues(
-					"NodePool", klog.KRef("", s.nodeClaimTemplates[i].NodePoolName),
+					"NodePool", klog.KRef("", s.nodeClaimTemplates[nodeClaimTemplateIndex].NodePoolName),
 				).Info(fmt.Sprintf(
 					"%d out of %d instance types were excluded because they would breach limits",
-					len(s.nodeClaimTemplates[i].InstanceTypeOptions)-len(its),
-					len(s.nodeClaimTemplates[i].InstanceTypeOptions),
+					len(s.nodeClaimTemplates[nodeClaimTemplateIndex].InstanceTypeOptions)-len(its),
+					len(s.nodeClaimTemplates[nodeClaimTemplateIndex].InstanceTypeOptions),
 				))
 			}
 		}
-		nodeClaim := NewNodeClaim(s.nodeClaimTemplates[i], s.topology, s.daemonOverhead[s.nodeClaimTemplates[i]], s.daemonHostPortUsage[s.nodeClaimTemplates[i]], its, s.reservationManager, s.reservedOfferingMode)
-		r, its, ofs, err := nodeClaim.CanAdd(ctx, pod, s.cachedPodData[pod.UID])
+		nodeClaim := NewNodeClaim(s.nodeClaimTemplates[nodeClaimTemplateIndex], s.topology, s.daemonOverhead[s.nodeClaimTemplates[nodeClaimTemplateIndex]], s.daemonHostPortUsage[s.nodeClaimTemplates[nodeClaimTemplateIndex]], its, s.reservationManager, s.reservedOfferingMode)
```

Comment:
I think we had mentioned that you should relax within a NodePool first before you try the next one -- I actually think that this makes the implementation easier -- it means that you don't have to iterate through the NodePools twice and can just try and get as high of a minValues as possible within the instance type filtering